server/src/main/java/io/crate/user/metadata/RolesMetadata.java

```diff
@@ -0,0 +1,247 @@
+     *   },
+     *   ...
+     * }
+     */
+    public static RolesMetadata fromXContent(XContentParser parser) throws IOException {
+        Map<String, Role> roles = new HashMap<>();
+        XContentParser.Token token = parser.nextToken();
+
+        if (token == XContentParser.Token.FIELD_NAME && parser.currentName().equals(TYPE)) {
+            token = parser.nextToken();
+            if (token == XContentParser.Token.START_OBJECT) {
+                while (parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+                    String roleName = parser.currentName();
+                    boolean isUser = false;
+                    SecureHash secureHash = null;
+                    if (parser.nextToken() == XContentParser.Token.START_OBJECT) {
+                        while (secureHash == null && parser.nextToken() == XContentParser.Token.FIELD_NAME) {
+                            switch (parser.currentName()) {
+                                case "is_user":
+                                    parser.nextToken();
+                                    isUser = parser.booleanValue();
+                                    break;
+                                case "secure_hash":
+                                    secureHash = SecureHash.fromXContent(parser);
+                                    break;
+                                default:
+                                    throw new ElasticsearchParseException("failed to parse roles, unexpected token: " +
+                                        parser.currentToken());
+                            }
+                        }
+                        if (isUser) {
+                            roles.put(roleName, Role.userOf(roleName, secureHash));
+                        } else {
+                            roles.put(roleName, Role.roleOf(roleName));
+                        }
+                    } else {
+                        // each custom metadata is packed inside an object.
+                        throw new ElasticsearchParseException("failed to parse roles, expected an object token at start");
+                    }
+                }
+            } else {
+                // each custom metadata is packed inside an object.
+                throw new ElasticsearchParseException("failed to parse roles, expected an object token at start");
+            }
+            if (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+                // each custom metadata is packed inside an object.
+                // each custom must move the parser to the end otherwise possible following customs won't be read
+                throw new ElasticsearchParseException("failed to parse roles, expected an object token at the end");
+            }
+        }
+        return new RolesMetadata(roles);
+    }
+
+    public static RolesMetadata rolesMetadataFromClusterState(Metadata.Builder mdBuilder, UsersMetadata oldUsersMetadata, RolesMetadata oldRolesMetadata) {
```

Comment:
@seut How about a `newRolesMetadata()` ? I'm also fine with simple `of()`.

Reply:
yep, agree, `of()` it is, I also removed the static import to make it more visible when used:
`RolesMetadata.of()`